### PR TITLE
Add initial scaffolding for default bucket (take 2)

### PR DESCRIPTION
**Pull request by sammcveety**
@tgroh, could you please take a look?

- Provide a hook for determining the project number associated with a bucket
- Provide an API for looking up the project number from a project ID

Re-opened because the other PR wasn't clearing the extra commit history, despite the branch being correct.

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcpProjectUtil.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+      FluentBackoff.DEFAULT.withMaxRetries(3).withInitialBackoff(Duration.millis(200));
+
+  /** Client for the CRM API. */
+  private CloudResourceManager crmClient;
+
+  private GcpProjectUtil(CloudResourceManager crmClient) {
+    this.crmClient = crmClient;
+  }
+
+  // Use this only for testing purposes.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcpProjectUtil.java]: mark `@VisibleForTesting`

Can this be package-private?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java [hunk 2]
```diff
@@ -349,16 +352,48 @@ public boolean bucketExists(GcsPath path) throws IOException {
...
+   * propagates an exception.
+   */
+  public void createBucket(String projectId, Bucket bucket) throws IOException {
+    createBucket(
+        projectId, bucket, BACKOFF_FACTORY.backoff(), Sleeper.DEFAULT);
+  }
+
+  /**
    * Returns whether the GCS bucket exists. This will return false if the bucket
    * is inaccessible due to permissions.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java]: This seems like a really strange contract. Could this be renamed to something like `bucketAccessible`?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/GcsUtilTest.java [hunk 3]
```diff
@@ -380,6 +382,56 @@ public void testRetryFileSize() throws IOException {
...
+    Storage mockStorage = Mockito.mock(Storage.class);
+    gcsUtil.setStorageClient(mockStorage);
+
+    Storage.Buckets mockStorageObjects = Mockito.mock(Storage.Buckets.class);
+    Storage.Buckets.Insert mockStorageInsert = Mockito.mock(Storage.Buckets.Insert.class);
+
+    BackOff mockBackOff = FluentBackoff.DEFAULT.backoff();
+    GoogleJsonResponseException expectedException =
+        googleJsonResponseException(HttpStatusCodes.STATUS_CODE_FORBIDDEN,
+            "Waves hand mysteriously", "These aren't the buckets your looking for");
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/util/GcsUtilTest.java]: you're

(My god I'm thrilled to be able to leave this comment)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcpProjectUtil.java (hunk 1 above)
Comment (sammcveety) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcpProjectUtil.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java (hunk 2 above)
Comment (sammcveety) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java]: I can change the hidden one, but the other one is part of the public API, so I think it has to stay?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/GcsUtilTest.java (hunk 3 above)
Comment (sammcveety) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/util/GcsUtilTest.java]: Argh, copy paste fail.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java]: We can change it beam but not Dataflow. It's probably worth getting right here, though it doesn't need to be in this PR (in which case, please file a JIRA)

**Comment by sammcveety** [comment]
Method renamed, all ready to go.

**Comment by lukecwik** [comment]
Merging

**Comment by lukecwik** [comment]
Thanks for fixing the git history.
Compared this with a copy I made in lukecwik/incubator-beam:finish-pr-1014 and it looked the same.
Merging

**Pull request closed without merging**